### DISCUSSION TITLE
Need help with diagnosing proxy issues

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#516] [Everyday usage]
Hi,

When I work from home, I connect my laptop to a VPN. My laptop is also pre-configured to use Automatic Proxy Configuration. In Terminal I use [Alpaca](https://github.com/samuong/alpaca) to simplify proxy settings.

Anyway, when I am connected to the proxy and the VPN, I can't update home-brew unless I disconnect from the VPN and turn off proxy which Alpaca can do transparently. From its README:

> When moving from, say, a corporate network to a public WiFi network (or vice-versa), the proxies listed in the PAC script might become unreachable. When this happens, Alpaca will temporarily bypass the parent proxy and send requests directly, so there's no need to manually unset/re-set http_proxy and https_proxy as you move between networks.

Anyway, here's an example error message I get when I am connected to our VPN:

```
brew update --verbose
Checking if we need to fetch /usr/local/Homebrew...
Checking if we need to fetch /usr/local/Homebrew/Library/Taps/anz-bank/homebrew-sysl...
Checking if we need to fetch /usr/local/Homebrew/Library/Taps/golangci/homebrew-tap...
Checking if we need to fetch /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask...
Checking if we need to fetch /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core...
Checking if we need to fetch /usr/local/Homebrew/Library/Taps/samuong/homebrew-alpaca...
Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core...
Fetching /usr/local/Homebrew...
Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask...
remote: Enumerating objects: 138, done.
remote: Counting objects: 100% (138/138), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 170 (delta 131), reused 134 (delta 131), pack-reused 32
Receiving objects: 100% (170/170), 47.44 KiB | 272.00 KiB/s, done.
Resolving deltas: 100% (134/134), completed with 69 local objects.
From https://github.com/Homebrew/brew
   b29c27b9a..99a7e2192  master     -> origin/master
❌ fatal: unable to access 'https://github.com/Homebrew/homebrew-core/': Malformed encoding found in chunked-encoding
❌ fatal: unable to access 'https://github.com/Homebrew/homebrew-cask/': Malformed encoding found in chunked-encoding
❌ Error: Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core failed!
❌ Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask failed!
Updating /usr/local/Homebrew...
First, rewinding head to replay your work on top of it...
Fast-forwarded master to origin/master.

Updated Homebrew from b29c27b9a to 99a7e2192.
No changes to formulae.
```

I want to find the root cause of this error and fix it. Any advice on how to proceed?

```
$ brew config
HOMEBREW_VERSION: 2.7.3-27-g99a7e21
ORIGIN: https://github.com/Homebrew/brew
HEAD: 99a7e2192c8ad825a9b5fdc6bee65865de1b9798
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 4a945f58e807c632227561da9256dd5429e41dfd
Core tap last commit: 26 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
http_proxy: http://localhost:3128
https_proxy: http://localhost:3128
no_proxy: <redacted>
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.32.28
Xcode: N/A
```

Here's a screen recording that shows the error even when I just clone the home-brew repo:

https://user-images.githubusercontent.com/75105145/104392856-225ee880-5597-11eb-9fd4-17df5aa366d3.mp4

**Comment by gromgit** [comment]
This error:
```
❌ fatal: unable to access 'https://github.com/Homebrew/homebrew-core/': Malformed encoding found in chunked-encoding
```
says `git` was able to connect to the GitHub servers, but got responses that were mangled by the proxy and/or Alpaca (which is itself a proxy).

Your next step should be to raise an issue upstream (i.e. Alpaca), as this is not something that Homebrew can help solve. Alternatively, unset your proxy environment variables when using Homebrew, or anything else that uses Git.

_Sidenote:_ Screen recordings have all the disadvantages of screenshots, and also prevent folks from seeing the big picture at a go (you have to zip back and forth to mentally compare screen images, etc.) Graphical representations of text output are a universal pain to deal with, so please don't do that. When asking for tech help anywhere, **let text be text** (i.e. dump the output to a text file, and post that file as a gist or pastebin if it's too large to post inline).

**Reply by ghost** [reply]
Thanks. I can't unset the proxy env vars as the only way to access external resources when on VPN is through the proxy. And if I disconnect the VPN, my desktop apps go nuts and can't connect to the internet and spawn a dozen dialog boxes.

Regarding:

> fatal: unable to access 'https://github.com/Homebrew/homebrew-core/': Malformed encoding found in chunked-encoding

Is there a way to configure git to log the http response with the malformed encoding? Or should I use Wireshark to see what's going on?

**Reply by jonchang** [reply]
GIT_CURL_VERBOSE and friends might be able to tell you what's going on.
